lib/chef/knife/vsphere_vm_snapshot.rb

```diff
@@ -147,17 +147,23 @@ def find_node(tree, name)
   end
 
   def display_node(node, current, shift = 1)
-    descr = node.name + ' ' + node.createTime.iso8601
-    out = ''
-    out << '+--' * shift
+    snapshot_tree = {}
+    children = []
+    unless node.childSnapshotList.empty?
+      node.childSnapshotList.each { |item| children << display_node(item, current, shift + 1) }
```

Comment:
This might benefit from using `map` instead of `each` then you could assign directly to `children` instead of populating it from the inside.

Also since you can call `#empty?` on it chances are the `#map` would return an empty array anyway so you wouldn't need the outer `unless`.

Reply:
Thanks for asking the question! In Ruby (and other functional languages) a _map_ means to take a collection of objects and apply a function to each of them, then return that new collection.

The usual example of `map` is to do some math on an array of numbers:

```Ruby
irb(main):004:0> [1, 2, 3].map { |i| i * i }
=> [1, 4, 9]
```

Which you _could_ have written as 
```Ruby
irb(main):005:0> temp = []
=> []
irb(main):006:0> [1, 2, 3].each { |i| temp << i * i }
=> [1, 2, 3]
irb(main):007:0> temp
=> [1, 4, 9]
```

So whenever you see

```Ruby
something = []
original.each { |item| something << somefunction(item) }
# use something
```
what you're doing is a map operation. Ruby let's us shorten that to

```Ruby
something = original.map { |item| somefunction(item) }
```

which gives you two benefits. The first is a small one, in that you can start chaining these operations. The second is that it clearly communicates what's going on. The important part is that you're assigning something, in your case, `children`, so this moves it out to the left where people expect it. And without the assignment buried in the block it becomes clear that the operation you're doing on the snapshot list is finding the sub nodes.

From there, we can look at the guard clause `unless node.childSnapshotList.empty?`. If the snapshot list happens to be empty, `children` will keep its initialized value of `[]`. However the result of iterating over an empty array would be to never run the block, or in the case of `map`, to return an empty array:

```Ruby
irb(main):002:0> [].map { |item| item * 2 }
=> []
```

So while it can be nice to have a guard clause to indicate some behaviour we want to check for, having no snapshots isn't really a problem because iterating over an empty snapshot list will give us an empty array.

Which reduces the code to

`children = node.childSnapshotList.map { |item| display_node(item, current, shift + 1) }`

No need to initialize `children` either. Someone reading the code can pretty clearly see that we're building a list of children based on running `display_node` on each childSnapshot.

Reply:
ah now I understand...thanks for this very useful info. .will definitely keep it in mind when I'm using Ruby...that's very cool!...I'll make that change :)